phys2bids/physio_obj.py

```diff
@@ -0,0 +1,403 @@
+        # comparing it to a given threshold.
+        trigger_deriv = np.diff(cls.timeseries[1])
+        timepoints = trigger_deriv > thr
+        num_timepoints_found = timepoints.sum()
+        time_offset = cls.timeseries[0][timepoints.argmax()]
+
+        if num_timepoints_expected:
+            print('Checking number of timepoints')
+            if num_timepoints_found > num_timepoints_expected:
+                timepoints_extra = (num_timepoints_found -
+                                    num_timepoints_expected)
+                print(f'Found {timepoints_extra} timepoints'
+                      'more than expected!\n'
+                      'Assuming extra timepoints are at the end '
+                      '(try again with a more conservative thr)')
+
+            elif num_timepoints_found < num_timepoints_expected:
+                timepoints_missing = (num_timepoints_expected -
+                                      num_timepoints_found)
+                print(f'Found {timepoints_missing} timepoints'
+                      'less than expected!')
+                if tr:
+                    print('Correcting time offset, assuming missing timepoints'
+                          ' are at the beginning (try again with '
+                          'a more liberal thr')
+                    time_offset -= (timepoints_missing * tr)
+                else:
+                    print('Can\'t correct time offset, (try again specifying '
+                          'tr or with a more liberal thr')
+
+            else:
+                print('Found just the right amount of timepoints!')
+
+        else:
+            print('Cannot check the number of timepoints')
+
+        cls.timeseries[0] -= time_offset
+        cls.num_timepoints_found = num_timepoints_found
+
+
+class blueprint_output():
+    """
+    Main output object for phys2bids.
+    Contains the blueprint to be exported.
+
+    Properties - Input
+    ------------------
+    timeseries : (ch x tps) :obj:`numpy.ndarray`
+        Numpy 2d array of timeseries
+        Contains all the timeseries recorded.
+        Impose same frequency!
+    freq : float
+        Shared frequency of the object.
+            Properties
+    ch_name : (ch) list of strings
+        List of names of the channels - can be the header of the columns
+        in the output files.
+    units : (ch) list of strings
+        List of the units of the channels.
+    start_time : float
+        Starting time of acquisition (equivalent to first TR,
+        or to the opposite sign of the time offset).
+
+    Methods
+    -------
+    return_index:
+        Returns the proper list entry of all the
+        properties of the object, given an index.
+    delete_at_index:
+        Returns all the proper list entry of the
+        properties of the object, given an index.
+    init_from_blueprint:
+        method to populate from input blueprint instead of init
```

Comment:
I'm not sure I'm getting what benefits this has over `blueprint_input`. The main differences (from what I can tell!) are that `timeseries` is required to be an array instead of a list of arrays, and this doesn't have a few of the methods that the `blueprint_input` does (e.g., `.check_trigger_amount()` and `.rename_channels()`). Could you clarify what purpose this class is serving over-and-above `blueprint_input`?

Reply:
So here's the thing. For latter uses (e.g. if we go on with preprocessing or data analysis), I'd say numpy arrays are a bit easier to work with, rather than list of numpy arrays.
However, as the input might have different frequencies, we need lists at the beginning.
Also, in one frequencies are a list and in the other it's a float, and we need to really check this and be sure it's respected.

What would be a solution otherwise?